src/Agent.Worker/ResourceMetricsManager.cs

```diff
@@ -38,78 +47,329 @@ public void Setup(IExecutionContext context)
+            var processStartInfoOutput = "";
+
+            if (PlatformUtil.RunningOnWindows)
+            {
+                processStartInfo.FileName = "wmic";
+                processStartInfo.Arguments = "OS GET FreePhysicalMemory,TotalVisibleMemorySize /Value";
+                processStartInfo.RedirectStandardOutput = true;
+
+                using (var process = Process.Start(processStartInfo))
+                {
+                    processStartInfoOutput = process.StandardOutput.ReadToEnd();
+                }
+
+                var processStartInfoOutputString = processStartInfoOutput.Trim().Split("\n");
+
+                var freeMemory = Int32.Parse(processStartInfoOutputString[0].Split("=", StringSplitOptions.RemoveEmptyEntries)[1]);
+                var totalMemory = Int32.Parse(processStartInfoOutputString[1].Split("=", StringSplitOptions.RemoveEmptyEntries)[1]);
+
+                memoryInfo.TotalMemoryMB = totalMemory / 1024;
+                memoryInfo.UsedMemoryMB = (totalMemory - freeMemory) / 1024;
             }
+
+            if (PlatformUtil.RunningOnLinux)
+            {
+                try
+                {
+                    processStartInfo.FileName = "free";
+                    processStartInfo.Arguments = "-m";
+                    processStartInfo.RedirectStandardOutput = true;
+
+                    using (var process = Process.Start(processStartInfo))
+                    {
+                        processStartInfoOutput = process.StandardOutput.ReadToEnd();
+                    }
+
+                    var processStartInfoOutputString = processStartInfoOutput.Split("\n");
+                    var memoryInfoString = processStartInfoOutputString[1].Split(" ", StringSplitOptions.RemoveEmptyEntries);
+
+                    if (memoryInfoString.Length != 7)
+                    {
+                        throw new MemoryMonitoringUtilityIsNotAvaliableException("Utility has non-default output");
+                    }
+
+                    memoryInfo.TotalMemoryMB = Int32.Parse(memoryInfoString[1]);
+                    memoryInfo.UsedMemoryMB = Int32.Parse(memoryInfoString[2]);
+                }
+                catch (Win32Exception e)
+                {
+                    throw new MemoryMonitoringUtilityIsNotAvaliableException(e.Message);
+                }
+            }
+
+            if (PlatformUtil.RunningOnMacOS)
+            {
+                processStartInfo.FileName = "vm_stat";
+                processStartInfo.RedirectStandardOutput = true;
+
+                using (var process = Process.Start(processStartInfo))
+                {
+                    processStartInfoOutput = process.StandardOutput.ReadToEnd();
+                }
+
+                var processStartInfoOutputString = processStartInfoOutput.Split("\n");
+
+                var pageSize = Int32.Parse(processStartInfoOutputString[0].Split(" ", StringSplitOptions.RemoveEmptyEntries)[7]);
```

Comment:
1. no checks that the array indexes are in range
2. use TryParse

Reply:
Index out of range and other exceptions will be catched in `RunMemoryUtilizationMonitor()` and `GetMemoryInfoString()` methods which calls `GetMemoryInfo`, additional checks on MacOS seem unnecessary here, since we already safely handle everything in caller method - the checks for Linux was added because the user may not have a memory stat utility or may customize it output, and we want to log this exception in traces without explicit ADO warnings, but for `vm_stat` on MacOS and `WMIC` on Windows it seems not possible.